### PR TITLE
Added RedBlack BST and Sentinel Search

### LINKED ISSUE #436: [Feature Request] Searching Technique and BST
### Is your feature request related to a problem? Please describe.
Add Sentinel Linear Search in searching technique
Add Red Black Tree in binary search tree

### Describe the solution you'd like
Sentinel Linear Search is an optimized version of the linear search. It places a sentinel value (usually the search key) at the end of the array. This helps avoid repeated boundary checks while searching, improving efficiency. The sentinel simplifies the loop logic by ensuring that the search always finds a match (either the actual match or the sentinel), which is later verified.

- How it works:
  - Append the search element (sentinel) at the end of the array.
  - Loop through the array until the search element is found.
  - After exiting the loop, check if the element is within the original array bounds or if it’s the sentinel.

Red-Black Tree is a type of self-balancing binary search tree that guarantees O(log n) time complexity for search, insertion, and deletion operations. Red-Black Trees ensure that the tree remains balanced, improving the efficiency of these operations.

- Key properties:
  - Every node is either red or black.
  - The root is always black.
  - Red nodes cannot have red children (no two consecutive red nodes).
  - Every path from a node to its descendant leaves contains the same number of black nodes.

### Describe alternatives you've considered
A proper description of both the topic , how these works, algorithm and working code with output.

### Additional context
none

**Would you like to work on this feature?**  
- [x] Yes  
- [ ] No  

@ajay-dhangar kindly assign me this issue

### PR DESCRIPTION
**Pull request by purnima2904**
### Description
In the Binary Search Tree i have added about Red Black Tree explaining in detail its application , advantages and disadvantages , code , rules and properties
In the Searching Algorithm i have added about Sentinel Search explaining in detail about its use case , code, complexity , definition and characteristics

Fixes #436 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by purnima2904** [comment]
Hello @ajay-dhangar kindly review this PR.

File: docs/binary-search-tree/RedBlackTree.md [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+- **Root Property:** The root of the tree is always black.
+- **Red Property:** Red nodes cannot have red children (no two consecutive red nodes on any path).
+- **Black Property:** Every path from a node to its descendant null nodes (leaves) has the same number of black nodes.
+- **Leaf Property:** All leaves (NIL nodes) are black.
+
+These properties ensure that the longest path from the root to any leaf is no more than twice as long as the shortest path, maintaining the tree’s balance and efficient performance.
+
+## Rules for constructing Red-Black Tree:
+- The black height of the red-black tree is the number of black nodes on a path from the root node to a leaf node. 
+- Leaf nodes are also counted as black nodes. So, a red-black tree of height h has black height >= h/2.
```
Comment (ajay-dhangar) [inline docs/binary-search-tree/RedBlackTree.md]: replace `>` with `&gt;`

File: docs/binary-search-tree/RedBlackTree.md [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+- Height of a red-black tree with n nodes is h<= 2 log2(n + 1).
```
Comment (ajay-dhangar) [inline docs/binary-search-tree/RedBlackTree.md]: replace `<` with `&lt;`

File: docs/binary-search-tree/RedBlackTree.md [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+- All leaves (NIL) are black.
+- The black depth of a node is defined as the number of black nodes from the root to that node i.e the number of black ancestors.
+
+## Correct Red-Black Tree:
+
+    10B
+   /   \
+  5R    20B
+ / \    /  \
+2B  8B 15R 25B
```
Comment (ajay-dhangar) [inline docs/binary-search-tree/RedBlackTree.md]: replace 

```md
    10B
   /   \
  5R    20B
 / \    /  \
2B  8B 15R 25B
```

with

````md
```
    10B
   /   \
  5R    20B
 / \    /  \
2B  8B 15R 25B
```
````

**OR**

if possible then try to use the mermaid graph (https://mermaid.js.org/)

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
update your code based on suggestions

**Pull request closed without merging**